### ISSUE TITLE
Mucsub subscription messages are not receiving when create a new group in a other xmpp connection. 

### ISSUE THREAD

**Issue by vietnguyen87** [processone/ejabberd#4206]
### Environment 
&nbsp; 

ejabberd version: 19.5.0
OS: Ubuntu 18.04.2 LTS
Using the ejabberd/ecs docker container

**Comment by vietnguyen87** [comment]
@prefiks, @badlop Please help.!

**Comment by prefiks** [comment]
So in other words you created room with `create_room_with_opts` with subscribers and then when you sent message to a room using command (what exactly command?), that didn't get delivered to subscriber from create command, yes? I am just trying to see what i need to do to reproduce it.

**Comment by vietnguyen87** [comment]
yup, @prefiks i used two ways: 
1. I use admin api https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#send-message 
2. I use XMPP send stanza 
```
<message to='room3@conference.localhost' type='groupchat' id='e23ff0a35b4e9150' xml:lang='en'><body>testing </body></message>
``` 
<img width="1092" alt="image" src="https://github.com/processone/ejabberd/assets/3839506/b15e8bb9-7462-4544-9720-c2f275edbc2e">

**Comment by prefiks** [comment]
It's working in my tests, you are aware that mucsub will not send regular groupchat message, but event with wrapped message like this:
```xml
<message to='prefiks@localhost' from='sub2@conference.localhost' id='1713861977424826' xmlns='jabber:client'>
  <event xmlns='http://jabber.org/protocol/pubsub#event'>
    <items node='urn:xmpp:mucsub:nodes:messages'>
      <item id='1713861977424826'>
        <message to='prefiks@localhost' from='sub2@conference.localhost/p2' type='groupchat' id='10483677060557293314' xmlns='jabber:client'><body>test</body>
        </message>
      </item>
    </items>
  </event>
</message>
```

**Comment by vietnguyen87** [comment]
hi @prefiks, i call Admin Api send-stanza https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#send-stanza

```
<message to='7214252@localhost' from='room3@conference.localhost' id='1713865884466642' xmlns='jabber:client'><event xmlns='http://jabber.org/protocol/pubsub#event'><items node='urn:xmpp:mucsub:nodes:messages'><item id='1713865884466642'><message to='7214252@localhost' from='room3@conference.localhost/7238854' type='groupchat' id='1713865884466642' xmlns='jabber:client'><body>test</body></message></item></items></event></message>
```
but it did not work. I can't receive the message at SSE connection. Data has stored at db. 

![image](https://github.com/processone/ejabberd/assets/3839506/789f83c5-52ca-4e85-8ac9-078cb2fe5e69)

**Comment by prefiks** [comment]
So that message that i showed in previous message was a result of calling
```
ejabberdctl send_message groupchat user@localhost sub2@conference.localhost "" test
```
and that resulted in sending that event to subscribers in a room.

I think ejabberd sends event like this too, but your SSE server doesn't recognize it. Are you able to see all traffic that your code receive?

**Comment by vietnguyen87** [comment]
hi @prefiks, i can see the traffic come ejabberd server. I think SSE server doesn't recognize it because it's not auto join when a new group created. When i reconnect SSE and join all group that user subscribe then message receive.

**Comment by vietnguyen87** [comment]
Hi @prefiks, 

Even if I reuse the xmpp client id in the call flow through the **Chat APIs Server** to join the room i created. 

![image](https://github.com/processone/ejabberd/assets/3839506/6083e3a8-154e-499a-a095-b368a0410b18)

I also did not receive the message when I sent it.
```
send_message groupchat 7238854@localhost room4@conference.localhost "" TestBody
```
From admin website no any users join the room 
![image](https://github.com/processone/ejabberd/assets/3839506/62e80d3b-9ae9-463e-96a4-7e7eb3b69a54)

**Comment by prefiks** [comment]
So about send_message not sending message to a room, user that you use in 'from' must be room subscriber or muc admin, only those two classes can send message without joining room first.

**Comment by vietnguyen87** [comment]
hi @prefiks, Are these users who subscribe to the room?

![image](https://github.com/processone/ejabberd/assets/3839506/01346191-8393-453c-806f-34c0d54b000d)

When i created the room. all users are become subscriber.

**Comment by prefiks** [comment]
Yes that list of room subscribers.

**Comment by vietnguyen87** [comment]
yeah, and i using **7238854@localhost** 7238854 is a room subscriber to sending message to the room. But all other subscribers don't receive messages.

**Comment by prefiks** [comment]
I don't know, do you have anything logged by server when you call that?

**Comment by vietnguyen87** [comment]
yes, when i call send_message 
``` 
send_message groupchat 7238854@localhost room3@conference.localhost "" "TestBody no join group"
```
and log by server 
![image](https://github.com/processone/ejabberd/assets/3839506/13cc98f9-8968-4836-b652-c66c5d43a0dd)

log from ejabberd 
![image](https://github.com/processone/ejabberd/assets/3839506/370eab87-fc95-4ba3-bda9-b6edc325c01f)

**Comment by vietnguyen87** [comment]
ok @prefiks, let's me check again. But. i can't receive history message if i don't join group. in case user disconnected, the SSE connection lost. Currently, when user reconnect SSE, i join all group user subscribe.

**Comment by prefiks** [comment]
Messages with event are stored in both user archive and offline storage (so they will be redelivered when user comes online). So those are two ways you can get older mesages.

**Comment by vietnguyen87** [comment]
in case user login on the other devices. I will get the messages history from api?

**Comment by vietnguyen87** [comment]
Hi @prefiks, I received data from ejabberd with type **xmpp.PubsubEvent**, parsing xml and received the message.
<img width="921" alt="image" src="https://github.com/processone/ejabberd/assets/3839506/21814eb0-0e48-477c-af70-5370170c0eb3">

```
{
  "Node": "urn:xmpp:mucsub:nodes:messages",
  "Items": [
    {
      "ID": "1713956108027223",
      "InnerXML": "PG1lc3NhZ2UgdG89JzcyMTQyNTJAbG9jYWxob3N0JyBmcm9tPSdyb29tMUBjb25mZXJlbmNlLmxvY2FsaG9zdC83MjM4ODU0JyB0eXBlPSdncm91cGNoYXQnIGlkPScxMjE4OTA1NjA4NjQ4ODIyMTgzMicgeG1sbnM9J2phYmJlcjpjbGllbnQnPjxhcmNoaXZlZCBieT0ncm9vbTFAY29uZmVyZW5jZS5sb2NhbGhvc3QnIGlkPScxNzEzOTU2MTA4MDI3MjIzJyB4bWxucz0ndXJuOnhtcHA6bWFtOnRtcCcvPjxzdGFuemEtaWQgYnk9J3Jvb20xQGNvbmZlcmVuY2UubG9jYWxob3N0JyBpZD0nMTcxMzk1NjEwODAyNzIyMycgeG1sbnM9J3Vybjp4bXBwOnNpZDowJy8+PGJvZHk+aGFoYWhhaGEgPzwvYm9keT48L21lc3NhZ2U+"
    }
  ]
}

```

InnerXML after parse 
```
<message to='7214252@localhost' from='room1@conference.localhost/7238854' type='groupchat' id='12189056086488221832' xmlns='jabber:client'><archived by='room1@conference.localhost' id='1713956108027223' xmlns='urn:xmpp:mam:tmp'/><stanza-id by='room1@conference.localhost' id='1713956108027223' xmlns='urn:xmpp:sid:0'/><body>hahahaha ?</body></message>
``` 
So what is the difference between muc and mucsub?

1. Mucsub, I can get the subscriber list to display, right?
2. For both modules, do I have to call the api to get message history when the user logs in on another device?

Please help me understand them clearly.!

**Comment by prefiks** [comment]
Regular muc: you must be online, and explicitly join room to receive messages, when you go offline, you automatically leave room and messages stop arrive.

Mucsub: after subscribing room you will get event for every message until you unsubscribe from room. Messages will also arrive when you are offline, so with mod_offline enabled, they will be delivered to you after you reconnect.

**Comment by vietnguyen87** [comment]
Yeah, I understand clearly.! 
Thanks so much @prefiks. You save my day.! :D

**Comment by vietnguyen87** [comment]
> Messages with event are stored in both user archive and offline storage (so they will be redelivered when user comes online). So those are two ways you can get older mesages.

Hi @prefiks, with MUCSUB when user online. If user don't join the room, will the message be automatically deliver? What is the best practice for returning history messages to users?

**Comment by vietnguyen87** [comment]
How to know if a subscriber is online in a room @prefiks ?

**Comment by badlop** [comment]
> with MUCSUB when user online. If user don't join the room, will the message be automatically deliver?

If user is subscribed to messages of that room, and is online, when messages are sent to the room they are delivered to the user.

> What is the best practice for returning history messages to users?

MAM

> How to know if a subscriber is online in a room ?

It depends, who wants to know it?

**Comment by Neustradamus** [comment]
@vietnguyen87: Like you speak about XEP-0045, a moment ago, I have done a ticket about an update XEP version support here:
- https://github.com/processone/xmpp/issues/91

**Comment by badlop** [comment]
@vietnguyen87 What's the status of this issue? I mean, were your doubts answered, and the problems solved? Or at least, did you find an alternative workaround for them?

**Comment by vietnguyen87** [comment]
@badlop  Thank you very much for your enthusiasm. My problem has been solved. Let's me close the ticket.
